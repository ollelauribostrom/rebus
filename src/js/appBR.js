import { clickReset } from './components/ResetButton';

export function loadButtons(resetIdiom) {
  clickReset(resetIdiom);
  clickbr();
  clicken();
}

// BUTTON [PT-BR] - LOAD BUTTON
export function clickbr() {
  // CONDITION: if init as [pt-br] rebus, change design and hide button [ptbr]
  // ELSE: hide button [english]
  checkIsPTBR();

  const e = document.getElementById('button-ptbr'); // get button [pt-br]

  // HOVER FUNCTION: cursor (hover) [pt-br]
  mouseHoverPTBR(e);

  // CLICK FUNCTION: click on button [pt-br]
  e.addEventListener('click', () => {
    localStorage.setItem('flagBR', 'true');
  });
  // FLAG: now [pt-br] rebus
}

// BUTTON [EN] - LOAD BUTTON
export function clicken() {
  const a = document.getElementById('button-en'); // get button [english]

  // HOVER FUNCTION: cursor (hover) [english]
  mouseHoverEN(a);

  // CLICK FUNCTION: click on button [english]
  a.addEventListener('click', () => {
    localStorage.setItem('flagBR', 'false');
  });
  // FLAG: now [english / original] rebus
}

function mouseHoverPTBR(e) {
  e.addEventListener('mouseover', () => {
    document.getElementById('button-text-id-br').classList.add('button-text-show');
  });
  e.addEventListener('mouseout', () => {
    document.getElementById('button-text-id-br').classList.remove('button-text-show');
  });
}

function mouseHoverEN(a) {
  a.addEventListener('mouseover', () => {
    document.getElementById('button-text-id-en').classList.add('button-text-show');
  });
  a.addEventListener('mouseout', () => {
    document.getElementById('button-text-id-en').classList.remove('button-text-show');
  });
}

export function checkResetPTBR() {
  let resetIdiom = 'reset-english';
  if (localStorage.getItem('flagBR') === 'true') resetIdiom = 'reset-ptbr';
  return resetIdiom;
}

export function checkIsPTBR() {
  if (localStorage.getItem('flagBR') === 'true') {
    return (
      document.querySelector('.root').classList.add('rootbr'),
      document.getElementById('button-ptbr').classList.add('button-spam')
    );
  }
  return document.getElementById('button-en').classList.add('button-spam');
}
