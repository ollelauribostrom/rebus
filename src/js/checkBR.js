export function checkResetPTBR(param) {
  let resetIdiom = param;
  if (localStorage.getItem('flagBR') === 'true') {
    resetIdiom = 'reset-ptbr';
  } else {
    resetIdiom = 'reset-english';
  }
  return resetIdiom;
}

export function checkIsPTBR() {
  if (localStorage.getItem('flagBR') === 'true') {
    document.querySelector('.root').classList.add('rootbr');
    document.getElementById('button-ptbr').classList.add('button-spam');
  } else {
    document.getElementById('button-en').classList.add('button-spam');
  }
}
