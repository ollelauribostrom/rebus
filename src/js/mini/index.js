export { createComponent } from './component';
export { render } from './render';
export { createStore } from './store';
