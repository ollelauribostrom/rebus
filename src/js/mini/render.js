export function isEventCallback(key, value) {
  return typeof value === 'function' && key.substr(0, 2) === 'on';
}

export function addListeners(component, $element) {
  Object.entries(component.props).forEach(([key, value]) => {
    if (isEventCallback(key, value)) {
      const event = key.substr(2).toLowerCase();
      $element.addEventListener(event, value);
    }
  });
}

export function renderChildren(component, $element, placeholder) {
  component.children
    .map(child => renderComponent(child, $element))
    .forEach(childElement => placeholder.parentNode.insertBefore(childElement, placeholder));
  placeholder.parentNode.removeChild(placeholder);
}

export function renderComponent(component, $parent) {
  const template = document.createElement('template');
  const rendered = component.render(component.props);
  template.innerHTML = rendered;
  const $element = template.content.firstElementChild;
  const childrenPlaceholder = $element.querySelector('children');
  if (childrenPlaceholder) {
    renderChildren(component, $element, childrenPlaceholder);
  }
  if (component.$element) {
    component.$parent.replaceChild($element, component.$element);
  }
  addListeners(component, $element);
  Object.assign(component, { $parent, $element, rendered });
  return $element;
}

export function callDidRender(component) {
  if (component.componentDidRender) {
    component.componentDidRender();
  }
  component.children.forEach(child => callDidRender(child));
}

export function render(component, root) {
  const element = renderComponent(component, root);
  root.appendChild(element);
  callDidRender(component);
}
