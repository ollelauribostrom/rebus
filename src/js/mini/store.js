export function createStore(initialState, actions) {
  const store = {
    connectedComponents: [],
    state: initialState
  };
  store.actions = new Proxy(actions, {
    get(target, key) {
      return async (...args) => {
        const newState = await actions[key](store.state, ...args);
        store.state = Object.assign(store.state, newState);
        store.connectedComponents.forEach(c => c.update());
      };
    }
  });
  store.connect = component => {
    const connected = new Proxy(component, {
      get(target, key) {
        if (key === 'render') {
          return (...args) => target.render(store.state, ...args);
        }
        return Reflect.get(target, key);
      }
    });
    store.connectedComponents.push(connected);
    return connected;
  };
  return store;
}
