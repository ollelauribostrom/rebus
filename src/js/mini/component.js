import { renderComponent } from './render';
import { isFunction } from './utils';

export function createComponent({
  props = {},
  children = [],
  render,
  componentDidMount,
  componentDidUpdate
}) {
  return {
    props,
    children,
    render,
    componentDidMount,
    componentDidUpdate,
    update(updateProps) {
      // this.updateProps = updateProps;
      if (this.rendered !== this.render(this.props)) {
        renderComponent(this, this.$parent);
        if (isFunction(this.componentDidUpdate)) {
          this.componentDidUpdate();
        }
      }
    }
  };
}
