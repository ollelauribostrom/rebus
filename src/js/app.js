import * as Sentry from '@sentry/browser';
import { render } from './mini';
import { App, Logo, GithubCorner, ChangeButton, Rebus } from './components';
import { actions } from './store';
import '../css/main.css';

export const onKeyUp = event => {
  switch (event.key) {
    case 'Left':
    case 'ArrowLeft':
      return actions.prev();

    case ' ':
    case 'Spacebar':
    case 'Right':
    case 'ArrowRight':
      return actions.next();

    default:
      return undefined;
  }
};

export function registerListeners() {
  document.addEventListener('keyup', onKeyUp);
}

export function init() {
  try {
    return render(
      App(
        Logo(),
        GithubCorner({ url: 'https://github.com/ollelauribostrom/rebus' }),
        ChangeButton({
          className: 'change-button--prev',
          onClick: () => actions.prev()
        }),
        Rebus({
          charInput: (input, wordIndex, charIndex) => {
            const confettiCanon = document.querySelector('.confetti-canon');
            actions.setInput(input, wordIndex, charIndex);
            actions.check(confettiCanon);
          }
        }),
        ChangeButton({
          className: 'change-button--next',
          onClick: () => actions.next()
        })
      ),
      document.querySelector('.root')
    );
  } catch (err) {
    return Sentry.captureException(err);
  }
}

if (!global || !global.isTestRun) {
  Sentry.init({ dsn: 'https://8f025bee12e84d9b8a16e9c3b9155ce8@sentry.io/1300214' });
  init();
  registerListeners();
}
