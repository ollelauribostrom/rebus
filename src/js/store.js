import { confetti } from 'dom-confetti';
import { createStore } from './mini';
import { getRebuses, markRebusAsAnswered } from './rebuses';

export const actionsCreators = {
  next: ({ current, rebuses }) => ({
    current: current < rebuses.length - 1 ? current + 1 : 0,
    animation: 'flip-vertical-right',
    incorrectAnswerCount: 0
  }),
  prev: ({ current, rebuses }) => ({
    current: current > 0 ? current - 1 : rebuses.length - 1,
    animation: 'flip-vertical-left',
    incorrectAnswerCount: 0
  }),
  setCurrent: ({ rebuses }, id) => {
    const index = rebuses.findIndex(rebus => rebus.id === id);
    if (index > 0) {
      return { current: index };
    }
    return {};
  },
  setInput: ({ current, rebuses }, input, wordIndex, charIndex) => {
    const rebus = rebuses[current];
    const previousWords = rebus.words.slice(0, wordIndex).join('');
    const index = wordIndex > 0 ? previousWords.length + charIndex : charIndex;
    const updatedRebuses = [...rebuses];
    updatedRebuses[current].input[index] = input;
    return { updatedRebuses };
  },
  check: ({ current, rebuses, incorrectAnswerCount }, confettiCanon) => {
    const rebus = rebuses[current];
    const input = rebus.input.join('').toUpperCase();
    const answer = rebus.words.join('').toUpperCase();
    if (input.length !== answer.length) {
      return {};
    }
    if (input === answer) {
      markRebusAsAnswered(rebus.id);
      confetti(confettiCanon);
      const updatedRebuses = [...rebuses];
      updatedRebuses[current].isAnswered = true;
      return { updatedRebuses, animation: 'none', incorrectAnswerCount: 0 };
    }
    return { incorrectAnswerCount: incorrectAnswerCount + 1 };
  },
  shake: ({ current, rebuses }) => {
    const rebus = rebuses[current];
    const isAnswered = rebus.words.join('').toUpperCase() === rebus.input.join('').toUpperCase();
    if (!isAnswered) {
      return { animation: 'shake' };
    }
    return {};
  }
};

export const initialState = {
  current: 0,
  animation: 'none',
  rebuses: getRebuses(),
  incorrectAnswerCount: 0
};

export const { connect, actions } = createStore(initialState, actionsCreators);
