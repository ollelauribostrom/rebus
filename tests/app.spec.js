import { init, registerListeners, onKeyUp } from '../src/js/app';
import { actions as actionsMock } from '../src/js/store';
import * as renderMock from '../src/js/mini/render';

jest.mock('../src/js/mini/render', () => {
  const mock = {};
  function reset() {
    Object.assign(mock, {
      render: jest.fn(),
      reset
    });
  }
  reset();
  return mock;
});
jest.mock('../src/js/store');

const oldAddEventListener = document.addEventListener;

jest.mock('../src/js/store', () => ({
  actions: {
    next: jest.fn(),
    prev: jest.fn()
  },
  connect: component => component
}));

afterEach(() => {
  renderMock.reset();
  document.addEventListener = jest.fn();
  actionsMock.prev.mockClear();
  actionsMock.next.mockClear();
});

afterAll(() => {
  document.addEventListener = oldAddEventListener;
});

const setup = () => {
  const container = document.createElement('div');
  container.className = 'root';
  document.body.append(container);
  return { container };
};

describe('Tests for app', () => {
  describe('init', () => {
    it('renders the app into the root element', () => {
      const { container } = setup();
      init();
      expect(renderMock.render).toHaveBeenCalledTimes(1);
      const firstCall = renderMock.render.mock.calls[0];
      const [firstArg, secondArg] = firstCall;
      expect(firstArg).toMatchSnapshot();
      expect(secondArg).toEqual(container);
    });
  });

  it('registers listeners', () => {
    registerListeners();
    expect(document.addEventListener).toHaveBeenCalledWith('keyup', onKeyUp);
  });

  describe('onKeyUp', () => {
    const testKeyPress = (key, action) => {
      it(`sends action on keypress with key ${key}`, () => {
        onKeyUp({ key });
        expect(action).toHaveBeenCalled();
      });
    };

    testKeyPress('Left', actionsMock.prev);
    testKeyPress('ArrowLeft', actionsMock.prev);

    testKeyPress('Right', actionsMock.next);
    testKeyPress('ArrowRight', actionsMock.next);

    testKeyPress(' ', actionsMock.next);
    testKeyPress('Spacebar', actionsMock.next);
  });
});
