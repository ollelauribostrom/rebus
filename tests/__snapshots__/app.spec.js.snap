// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for app init renders the app into the root element 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "props": Object {},
      "render": [Function],
      "update": [Function],
    },
    Object {
      "children": Array [],
      "props": Object {
        "url": "https://github.com/ollelauribostrom/rebus",
      },
      "render": [Function],
      "update": [Function],
    },
    Object {
      "children": Array [],
      "props": Object {
        "className": "change-button--prev",
        "onClick": [Function],
      },
      "render": [Function],
      "update": [Function],
    },
    Object {
      "children": Array [],
      "props": Object {
        "charInput": [Function],
      },
      "render": [Function],
      "update": [Function],
    },
    Object {
      "children": Array [],
      "props": Object {
        "className": "change-button--next",
        "onClick": [Function],
      },
      "render": [Function],
      "update": [Function],
    },
  ],
  "props": Object {},
  "render": [Function],
  "update": [Function],
}
`;
